### PR TITLE
Expose staking event hashes + add GetGenesisDocument to consensus client API

### PR DESCRIPTION
**Pull request by abukosek**

### CODE REVIEW

File: go/staking/api/api.go [hunk 1]
```diff
@@ -156,6 +157,7 @@ type EscrowEvent struct {
 
 // Event signifies a staking event, returned via GetEvents.
 type Event struct {
+	Hash          hash.Hash      `json:"hash,omitempty"`
```
Comment (kostko) [inline go/staking/api/api.go:160]:
```suggestion
	TxHash          hash.Hash      `json:"tx_hash,omitempty"`
	
```

File: go/consensus/tendermint/tendermint.go [hunk 2]
```diff
@@ -446,6 +446,24 @@ func (t *tendermintService) StateToGenesis(ctx context.Context, blockHeight int6
 	}, nil
 }
 
+func (t *tendermintService) GetGenesisDocument(ctx context.Context) (*genesisAPI.Document, error) {
+	genesisFileProvider, err := file.DefaultFileProvider()
```
Comment (kostko) [inline go/consensus/tendermint/tendermint.go:450]: Why not just return `t.genesis`?

Also I now see that `StateToGenesis` also incorrectly uses `file.DefaultFileProvider`. While we're here it should be fixed to use `t.genesis` instead.

File: go/consensus/tendermint/tendermint.go [hunk 3]
```diff
@@ -446,6 +446,24 @@ func (t *tendermintService) StateToGenesis(ctx context.Context, blockHeight int6
 	}, nil
 }
 
+func (t *tendermintService) GetGenesisDocument(ctx context.Context) (*genesisAPI.Document, error) {
```
Comment (kostko) [inline go/consensus/tendermint/tendermint.go:449]: Don't forget to remove the existing `GetGenesis` method that exists in the tendermint-specific service (any callers should use this one).

File: go/consensus/tendermint/staking/staking.go [hunk 4]
```diff
@@ -37,6 +38,14 @@ type tendermintBackend struct {
 	closedCh chan struct{}
 }
 
+// Extend the abci Event struct with the transaction hash if the event was
+// the result of a transaction.  Block events have Hash set to the empty hash.
+type abciEventWithHash struct {
+	abcitypes.Event
+
+	Hash hash.Hash
```
Comment (kostko) [inline go/consensus/tendermint/staking/staking.go:46]:
```suggestion
	TxHash hash.Hash
```

File: go/consensus/tendermint/tendermint.go (hunk 3 above)
Comment (Yawning) [reply go/consensus/tendermint/tendermint.go:449]: Is this the right thing to name the method?  It's a state dump, not a genesis document (#2886 is illustrative of the fact that there is a difference).

File: go/consensus/tendermint/tendermint.go (hunk 2 above)
Comment (abukosek) [reply go/consensus/tendermint/tendermint.go]: I lazily copied it from `StateToGenesis`, fixed both now :)

File: go/consensus/tendermint/tendermint.go (hunk 3 above)
Comment (abukosek) [reply go/consensus/tendermint/tendermint.go:441]: After the suggested changes, `GetGenesisDocument` returns `t.genesis`, so it should be equivalent to the old `GetGenesis` method.

**Comment by abukosek** [comment]
Thanks for the review comments, I've addressed them now, so this is ready for re-review :)

File: go/staking/api/api.go [hunk 5]
```diff
@@ -156,6 +157,7 @@ type EscrowEvent struct {
 
 // Event signifies a staking event, returned via GetEvents.
 type Event struct {
+	TxHash        hash.Hash      `json:"tx_hash,omitempty"`
```
Comment (kostko) [inline go/staking/api/api.go:160]: Add a blank line after TxHash to separate it from events.

**Comment by abukosek** [comment]
Done! :)

File: go/consensus/tendermint/staking/staking.go [hunk 6]
```diff
@@ -154,10 +180,34 @@ func (tb *tendermintBackend) GetEvents(ctx context.Context, height int64) ([]api
...
-		tmEvents = append(tmEvents, txResults.Events...)
+	tmEvents := convertTmBlockEvents(results.BeginBlockEvents, results.EndBlockEvents)
+	for txIdx, txResults := range results.TxsResults {
+		// The order of transactions in txns and results.TxsResults is
+		// supposed to match, so the same index in both slices refers to the
+		// same transaction.
+
+		// Generate hash of transaction.
+		var evHash hash.Hash
+		evHash.FromBytes(txns[txIdx])
```
Comment (kostko) [inline go/consensus/tendermint/staking/staking.go:202]: Use the freshly merged `hash.NewFromBytes` instead.

File: go/consensus/tendermint/staking/staking.go [hunk 7]
```diff
@@ -210,24 +260,36 @@ func (tb *tendermintBackend) worker(ctx context.Context) {
...
-	events = append(events, ev.ResultEndBlock.GetEvents()...)
+	events := convertTmBlockEvents(ev.ResultBeginBlock.GetEvents(), ev.ResultEndBlock.GetEvents())
 
 	_, _ = tb.onABCIEvents(ctx, events, ev.Block.Header.Height, true)
 }
 
 func (tb *tendermintBackend) onEventDataTx(ctx context.Context, tx tmtypes.EventDataTx) {
-	_, _ = tb.onABCIEvents(ctx, tx.Result.Events, tx.Height, true)
+	var evHash hash.Hash
+	evHash.FromBytes(tx.Tx)
```
Comment (kostko) [inline go/consensus/tendermint/staking/staking.go:270]: Use the freshly merged `hash.NewFromBytes` instead.

**Comment by abukosek** [comment]
Thanks!  Ready for re-review.

**Pull request merged**